### PR TITLE
[SPARK-35232][SQL] Nested column pruning should retain column metadata

### PR DESCRIPTION
**Pull request by sunchao**
### What changes were proposed in this pull request?

Retain column metadata during the process of nested column pruning, when constructing `StructField`. 

To test the above change, this also added the logic of column projection in `InMemoryTable`. Without the fix `DSV2CharVarcharDDLTestSuite` will fail.

### Why are the changes needed?

The column metadata is used in a few places such as re-constructing CHAR/VARCHAR information such as in [SPARK-33901](https://issues.apache.org/jira/browse/SPARK-33901). Therefore, we should retain the info during nested column pruning.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing tests.

### CODE REVIEW

**Comment by viirya** [comment]
> Retain column metadata during the process of nested column pruning, when constructing StructField.
To test the above change, this also added the logic of column projection in InMemoryTable. Without the fix DSV2CharVarcharDDLTestSuite will fail.

Does it mean if only applying `SchemaPruning` change without `InMemoryTable`, `DSV2CharVarcharDDLTestSuite` will fail?

**Comment by sunchao** [comment]
@viirya without the change in `InMemoryTable` the test will not fail. This is because `InMemoryTable` doesn't have column pruning at the moment and it just return the table schema as the read schema in `InMemoryBatchScan`, but a more realistic data source would use the `requestedSchema` (e.g., [Iceberg](https://github.com/apache/iceberg/blob/master/spark3/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java#L132)).

**Comment by sunchao** [comment]
Also it seems [SPARK-33901](https://issues.apache.org/jira/browse/SPARK-33901) (#30918) doesn't work for nested type. For instance, this test:
```scala
  test("SPARK-XXXXX: create table like should should not change table's schema (nested type)") {
    withTable( "tt") {
      sql(s"CREATE TABLE tt(s1 struct<i: CHAR(5), j: int>, s2 struct<c: VARCHAR(4), d: int>) " +
        s"USING $format")
      withView("t") {
        sql("CREATE VIEW t AS SELECT s1.i, s2.c FROM tt")
        checkTableSchemaTypeStr(Seq(Row("char(5)"), Row("varchar(4)")))
      }
    }
  }
```
doesn't work even with this fix. It seems this is related to how the metadata is handled: in nested schema the metadata is associated with the top-level field (e.g., `s1`, `s2`) instead of leaf nodes (e.g., `i`, `c`). cc @yaooqinn @cloud-fan

**Comment by sunchao** [comment]
kindly ping @viirya @cloud-fan @yaooqinn - it'd be great to get a review from you :)

**Comment by viirya** [comment]
> To test the above change, this also added the logic of column projection in InMemoryTable. Without the fix DSV2CharVarcharDDLTestSuite will fail.

Oh, you mean that by applying `InMemoryTable` related change here, `DSV2CharVarcharDDLTestSuite` will fail because nested column pruning doesn't retain metadata now. Right?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruning.scala
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruning.scala:131]: Can we add a unit test in `SchemaPruningSuite`? We can make `getRootFields` as `private[spark]` to test it.

**Comment by sunchao** [comment]
> Oh, you mean that by applying `InMemoryTable` related change here, `DSV2CharVarcharDDLTestSuite` will fail because nested column pruning doesn't retain metadata now. Right?

Yup exactly.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruning.scala
Comment (sunchao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruning.scala:131]: Good point. Let me add a test.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryTable.scala
Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryTable.scala:244]: `(metadataColumnNames ++ tableSchema.map(_.name))` can be out of `filter` loop?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryTable.scala
Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryTable.scala:527]: Is case-sensitivity a problem here?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryTable.scala
Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryTable.scala:543]: looks like we don't do char/varchar to string conversion before?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryTable.scala
Comment (sunchao) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryTable.scala:527]: Good question. Looking at `PushdownUtils.pruneColumns`, I see that we apply `SQLConf.resolver` when nested column pruning is enabled, but seems not so when it is disabled. IMO perhaps we should have better contract between Spark and data source implementors w.r.t `SupportsPushDownRequiredColumns.pruneColumns`, and Spark should guarantee that the `requiredSchema` passed in to the method should be a "subset" of the relation's schema (e.g., table schema).

File: sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryTable.scala
Comment (sunchao) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryTable.scala:543]: Yes I think this is not required actually. My bad.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryTable.scala
Comment (sunchao) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryTable.scala:527]: Oh actually, in both cases the `requiredSchema` are normalized according to case-sensitivity. This is done in `V2ScanRelationPushDown` (see `normalizedProjects` there.)

**Review by viirya — COMMENTED** [review]
Looks okay to me. Is the test failure related?

**Comment by sunchao** [comment]
Thanks. They don't seem related. I tested them locally and all passed.

**Comment by viirya** [comment]
retest this please

File: sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryTable.scala
Comment (viirya) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryTable.scala:532]: nit: we can use `row.isNullAt(index)`.

**Review by viirya — APPROVED** [review]
lgtm with one minor comment

**Comment by viirya** [comment]
I will go to merge this once CI passes.

**Comment by viirya** [comment]
Thanks @sunchao! Merging to master.

**Comment by sunchao** [comment]
Thanks @viirya for the review!

**Pull request closed without merging**
